File: apps/php-sdk/src/Models/AgentTraceEvent.php
<?php

declare(strict_types=1);

namespace Firecrawl\Models;

/**
 * A single agent trace event. One model covers every event `type`
 * (run.started, tool_call.finished, artifact.updated, ...); fields that do
 * not apply to a given type stay null.
 */
final class AgentTraceEvent
{
    public function __construct(
        private readonly ?string $type = null,
        private readonly ?int $schemaVersion = null,
        private readonly ?string $eventId = null,
        private readonly ?string $runId = null,
        private readonly ?string $occurredAt = null,
        private readonly ?int $producerSequence = null,
        private readonly ?AgentTraceAgentIdentity $agent = null,
        private readonly ?string $reason = null,
        private readonly ?string $outcome = null,
        private readonly ?AgentTraceError $error = null,
        private readonly ?int $durationMs = null,
        private readonly ?string $sessionId = null,
        private readonly ?string $phase = null,
        private readonly ?string $message = null,
        private readonly ?string $text = null,
        private readonly ?string $toolCallId = null,
        private readonly ?string $toolName = null,
        private readonly mixed $parameters = null,
        private readonly mixed $result = null,
        private readonly ?AgentTraceArtifactChange $artifact = null,
    ) {}

    /** @param array<string, mixed> $data */
    public static function fromArray(array $data): self
    {
        return new self(
            type: $data['type'] ?? null,
            schemaVersion: isset($data['schemaVersion']) ? (int) $data['schemaVersion'] : null,
            eventId: $data['eventId'] ?? null,
            runId: $data['runId'] ?? null,
            occurredAt: $data['occurredAt'] ?? null,
            producerSequence: isset($data['producerSequence']) ? (int) $data['producerSequence'] : null,
            agent: isset($data['agent']) && is_array($data['agent'])
                ? AgentTraceAgentIdentity::fromArray($data['agent'])
                : null,
            reason: $data['reason'] ?? null,
            outcome: $data['outcome'] ?? null,
            error: isset($data['error']) && is_array($data['error'])
                ? AgentTraceError::fromArray($data['error'])
                : null,
            durationMs: isset($data['durationMs']) ? (int) $data['durationMs'] : null,
            sessionId: $data['sessionId'] ?? null,
            phase: $data['phase'] ?? null,
            message: $data['message'] ?? null,
            text: $data['text'] ?? null,
            toolCallId: $data['toolCallId'] ?? null,
            toolName: $data['toolName'] ?? null,
            parameters: $data['parameters'] ?? null,
            result: $data['result'] ?? null,
            artifact: isset($data['artifact']) && is_array($data['artifact'])
                ? AgentTraceArtifactChange::fromArray($data['artifact'])
                : null,
        );
    }

    public function getType(): ?string
    {
        return $this->type;
    }

    public function getSchemaVersion(): ?int
    {
        return $this->schemaVersion;
    }

    public function getEventId(): ?string
    {
        return $this->eventId;
    }

    public function getRunId(): ?string
    {
        return $this->runId;
    }

    public function getOccurredAt(): ?string
    {
        return $this->occurredAt;
    }

    public function getProducerSequence(): ?int
    {
        return $this->producerSequence;
    }

    public function getAgent(): ?AgentTraceAgentIdentity
    {
        return $this->agent;
    }

    public function getReason(): ?string
    {
        return $this->reason;
    }

    public function getOutcome(): ?string
    {
        return $this->outcome;
    }

    public function getError(): ?AgentTraceError
    {
        return $this->error;
    }

    public function getDurationMs(): ?int
    {
        return $this->durationMs;
    }

    public function getSessionId(): ?string
    {
        return $this->sessionId;
    }

    public function getPhase(): ?string
    {
        return $this->phase;
    }

    public function getMessage(): ?string
    {
        return $this->message;
    }

    public function getText(): ?string
    {
        return $this->text;
    }

    public function getToolCallId(): ?string
    {
        return $this->toolCallId;
    }

    public function getToolName(): ?string
    {
        return $this->toolName;
    }

    public function getParameters(): mixed
    {
        return $this->parameters;
    }

    public function getResult(): mixed
    {
        return $this->result;
    }

    public function getArtifact(): ?AgentTraceArtifactChange
    {
        return $this->artifact;
    }
}

File: apps/java-sdk/src/main/java/com/firecrawl/models/AgentTraceEvent.java
package com.firecrawl.models;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

/**
 * A single event in an agent run trace.
 *
 * <p>Models all event kinds in one class; per-type fields are null when not
 * applicable. Known {@code type} values: {@code run.started},
 * {@code run.cancel_requested}, {@code run.finished}, {@code agent.started},
 * {@code agent.finished}, {@code browser.session.started},
 * {@code browser.session.finished}, {@code progress.reported},
 * {@code reasoning.summary}, {@code tool_call.started},
 * {@code tool_call.finished}, {@code artifact.updated}, {@code error.occurred}.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class AgentTraceEvent {

    private String type;

    // Base fields present on every event
    private int schemaVersion;
    private String eventId;
    private String runId;
    private String occurredAt;
    private long producerSequence;
    private AgentTraceAgentIdentity agent;

    // Per-type fields (nullable)
    private String reason;
    private String outcome;
    private AgentTraceError error;
    private Long durationMs;
    private String sessionId;
    private String phase;
    private String message;
    private String text;
    private String toolCallId;
    private String toolName;
    private Object parameters;
    private Object result;
    private AgentTraceArtifactChange artifact;

    public String getType() { return type; }
    public int getSchemaVersion() { return schemaVersion; }
    public String getEventId() { return eventId; }
    public String getRunId() { return runId; }
    public String getOccurredAt() { return occurredAt; }
    public long getProducerSequence() { return producerSequence; }
    public AgentTraceAgentIdentity getAgent() { return agent; }
    public String getReason() { return reason; }
    public String getOutcome() { return outcome; }
    public AgentTraceError getError() { return error; }
    public Long getDurationMs() { return durationMs; }
    public String getSessionId() { return sessionId; }
    public String getPhase() { return phase; }
    public String getMessage() { return message; }
    public String getText() { return text; }
    public String getToolCallId() { return toolCallId; }
    public String getToolName() { return toolName; }
    public Object getParameters() { return parameters; }
    public Object getResult() { return result; }
    public AgentTraceArtifactChange getArtifact() { return artifact; }

    @Override
    public String toString() {
        return "AgentTraceEvent{type=" + type + ", eventId=" + eventId + ", occurredAt=" + occurredAt + "}";
    }
}
